device/Microsoft.Azure.Devices.Client.Shared/ITokenRefresher.cs

```diff
@@ -0,0 +1,34 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.Devices.Client
+{
+    using System;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Interface to get a new token for the device or module
+    /// </summary>
+    interface ITokenRefresher
```

Comment:
Adding an interface is usually limiting the expansion of this type, especially if it's public.
If any kinds of additions are possible to this (and I believe there are), we should instead use `abstract class` which allows for default implementations.

Reply:
I am not sure I understand how adding the interface limits this. Especially when you think that the interface is internal only. 
We can use the Device* abstract class, but that is not what we have done for other AuthenticationMethods.

Reply:
From offline discussion: we can extract an `internal AuthenticateWithTokenRefresh` and move all common code within it exposing only the differences.
That way `ModuleAuthenticationWithTokenRefresh` won't have to publicly inherit from `DeviceAuthenticationWithTokenRefresh` (in case they diverge in public API in the future.